server-mirror-git/src/main/java/com/linecorp/centraldogma/server/internal/mirror/AbstractGitMirror.java

```diff
@@ -131,283 +136,285 @@ public GitMirror(Cron schedule, MirrorDirection direction, MirrorCredential cred
+            // the requested changes to build the new tree.
+            final DirCache dirCache = DirCache.newInCore();
+            final DirCacheBuilder builder = dirCache.builder();
+            builder.addTree(EMPTY_BYTE, 0, reader, headTreeId);
+            builder.finish();
+
+            try (ObjectInserter inserter = gitRepository.newObjectInserter()) {
+                addModifiedEntryToCache(localHead, dirCache, reader, inserter,
+                                        treeWalk, maxNumFiles, maxNumBytes);
+                // Add the mirror state file.
+                final MirrorState mirrorState = new MirrorState(localHead.text());
+                applyPathEdit(
+                        dirCache, new InsertText(mirrorStatePath.substring(1), // Strip the leading '/'.
+                                                 inserter,
+                                                 Jackson.writeValueAsPrettyString(mirrorState) + '\n'));
+            }
 
-        try (GitWithAuth git = openGit(workDir)) {
-            final String headBranchRefName = getHeadBranchRefName(git);
-            final ObjectId id = fetchRemoteHeadAndGetCommitId(git, headBranchRefName);
-            final Revision localRev = localRepo().normalizeNow(Revision.HEAD);
-
-            try (ObjectReader reader = git.getRepository().newObjectReader();
-                 TreeWalk treeWalk = new TreeWalk(reader);
-                 RevWalk revWalk = new RevWalk(reader)) {
-
-                // Prepare to traverse the tree.
-                treeWalk.addTree(revWalk.parseTree(id).getId());
-
-                // Check if local repository needs update.
-                final String mirrorStatePath = localPath() + MIRROR_STATE_FILE_NAME;
-                final Entry<?> mirrorState = localRepo().getOrNull(localRev, mirrorStatePath).join();
-                final String localSourceRevision;
-                if (mirrorState == null || mirrorState.type() != EntryType.JSON) {
-                    localSourceRevision = null;
-                } else {
-                    localSourceRevision = Jackson.treeToValue((TreeNode) mirrorState.content(),
-                                                              MirrorState.class).sourceRevision();
-                }
+            final ObjectId nextCommitId =
+                    commit(gitRepository, dirCache, headCommitId, localHead);
+            updateRef(gitRepository, revWalk, headBranchRefName, nextCommitId);
+        }
 
-                final String abbrId = reader.abbreviate(id).name();
-                if (id.name().equals(localSourceRevision)) {
-                    logger.info("Repository '{}' already at {}, {}#{}", localRepo().name(), abbrId,
-                                remoteRepoUri(), remoteBranch());
-                    return;
-                }
+        final PushCommand push = git.push();
+        configurator.accept(push);
```

Comment:
> So is it necessary that we always call configurator.accept for each command from now on?

Yes, 😉 

> I think I prefer the original style where Git automatically configures authentication for each command.

The reason why I've changed is to share the `client` and `session` for ls-remote, fetch and push commands.
Because the client and session must be closed after use, I think it's hard to use it with the old style.